Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection

the-chenergy opened the pull request:
#### e02e121abf243cc5c461af6def0d30df24eadf81
<pre>
Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection
<a href="https://rdar.apple.com/145795990">rdar://145795990</a>
<a href="https://bugs.webkit.org/show_bug.cgi?id=288776">https://bugs.webkit.org/show_bug.cgi?id=288776</a>

Reviewed by BJ Burg.

This work adds support to automatically inspect or pause the
ServiceWorkerDebuggableProxy, alongside minor refactoring of existing
code around auto-inspection.

The squashed commit includes three patches:

(Patch 1 of 3) Make pausing for automatic inspection a responsibility
of the RemoteInspectionTarget rather than RemoteInspector.

To summarize how automatic inspection generally works:
1. The web content marks its debuggable as inspectable.
2. The debuggable blocks its thread before starting code execution.
3. The auto-launched frontend, after initializing, notifies the backend.
4. The backend unblocks the debuggable to resume its code execution.

As the newly introduced ServiceWorkerDebuggableProxy lives in the UI
process but controls a different thread, during step 2, it&apos;s the service
worker thread that would need to be blocked, while the UI process
remains free to handle incoming messages from the remote inspector host.

This patch builds the ground work for supporting pausing in a different
thread by moving the handling of pausing or unpausing into the
debuggable instead of implicitly in RemoteInspector. With this change:
- From the RemoteInspector&apos;s perspective, it only needs to manage the
  set of debuggables waiting for a reply of auto-inspection policy.
  Notably, it now tells the debuggable to unpause when auto-inspection
  is rejected, rather than relying on the call to pauseWaitingForAutomaticInspection
  to return and assuming that it was a blocking operation.
- From the RemoteInspectionTarget&apos;s perspective, it should decide
  what the pausing behavior is, defaulting to blocking the current
  thread with a busy loop. The pause should then persist until an
  explicit call of unpauseForResolvedAutomaticInspection.

With this patch, it&apos;ll be easier to let ServiceWorkerDebuggableProxy
later define a custom pausing behavior that controls a different thread.

I tested on both Mac and iOS, with an app that spawns a JSContext, that
this change preserves normal and automatic inspection with no noticeable
difference from before.

* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp:
(Inspector::RemoteInspectionTarget::pauseWaitingForAutomaticInspection):
(Inspector::RemoteInspectionTarget::unpauseForResolvedAutomaticInspection):
(Inspector::RemoteInspectionTarget::unpauseForInitializedInspector): Deleted.
   - Use m_isPausedWaitingForAutomaticInspection to explicitly handle
     the period of pausing.
   - Rename unpauseForResolvedAutomaticInspection to indicate that this
     should be called in both the frontend-initialized case and, as of
     now, the rejected case as well.

* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteConnectionToTargetCocoa.mm:
(Inspector::RemoteConnectionToTarget::setup):
   - Make sure to call setupCompleted when we&apos;ve determined the answer
     of isAutomaticInspection to clean up that hash set entry, since
     it&apos;s no longer the debuggable&apos;s job to do this upon
     frontend-initialized.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspector.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
   - Rename m_automaticInspectionCandidates since &quot;pausing&quot; is no longer
     this object&apos;s responsibility, and this hash set is only here to
     know which incoming inspection requests were automatic.

(Inspector::RemoteInspector::waitingForAutomaticInspection): Deleted.
   - No longer used. Debuggables should not rely on this information
     to remain paused.

(Inspector::RemoteInspector::setupFailed):
(Inspector::RemoteInspector::setupCompleted):
(Inspector::RemoteInspector::setPendingMainThreadInitialization):
(Inspector::RemoteInspector::updateAutomaticInspectionCandidate):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::setupXPCConnectionIfNeeded):
(Inspector::RemoteInspector::xpcConnectionFailed):
(Inspector::RemoteInspector::receivedSetupMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionConfigurationMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionRejectMessage):
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::stopInternal):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::stopInternal):
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::frontendInitialized):
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.cpp:
(Inspector::JSGlobalObjectInspectorController::frontendInitialized):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - Adapt to renamed things.

---

(Patch 2 of 3) Move setup of ServiceWorkerDebuggable auto-inspection
into WebSWContextManagerConnection.

Implementing automatic inspection of the ServiceWorkerDebuggable
required marking the debuggable as inspectable and then handling when
to unpause. That work was originally done inside ServiceWorkerThreadProxy.
As we plan to implement the same function for ServiceWorkerDebuggableProxy,
which is created and largely controlled in WebSWContextManagerConnection,
it might be more readable and logical to have both implementations in
that class instead.

* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::registerServiceWorkerThreadForInstall):
(WebCore::SWContextManager::stopRunningDebuggerTasksOnServiceWorker):
* Source/WebCore/workers/service/context/SWContextManager.h:
(WebCore::SWContextManager::registerServiceWorkerThreadForInstall):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks): Deleted.
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
   - Move the work of handling the debuggable (steps 2 and 4) into WebSWContextManagerConnection
     where more implementation is anticipated.
   - Lay down a basic code structure for handling auto-inspection for
     when either the Debuggable or the DebuggableProxy is used.

* Source/WTF/wtf/PlatformEnableCocoa.h:
   - Add back the feature flag here as my work should end up providing
     support for both ServiceWorkerDebuggable and ServiceWorkerDebuggableProxy.

---

(Patch 3 of 3) Add support for auto-inspecting ServiceWorkerDebuggableProxy.

This patch adds support to automatically inspect or pause a
ServiceWorkerDebuggableProxy in a similar fashion to the same features
for JSGlobalObjectInspectorController and ServiceWorkerDebuggable.
The ServiceWorkerDebuggableProxy is unique in that, in step 2, it should
keep the UI process unblocked to handle incoming remote inspector
messages including the auto-inspection policy, while controlling when
to unblock the actual service worker thread. This is done by two IPC
messages:
- The async callback of CreateServiceWorkerDebuggable, which can
  indicate to unpause because there&apos;s no eligible remote inspector
  at all.
- The new message UnpauseServiceWorkerForRejectedAutomaticInspection can
  request unpause when the policy is received and it&apos;s a rejection.

These messages are from the UI process into the web process so they
should be relatively safe. If neither of the above two cases are met,
that means auto-inspection was accepted, and the web process will listen
to the frontendInitialized event to unblock the service worker thread.

I tested on Mac and iOS with a inspector host having Develop menu
on or off, or without support of auto-inspecting service workers and
didn&apos;t experience problems or hangs like what my last attempt introduced.

* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
(WebKit::WebSWContextManagerConnection::unpauseServiceWorkerForRejectedAutomaticInspection):
   - Step 1 and 2, creating and blocking the thread as needed.

(WebKit::WebSWContextManagerConnection::connectToInspector):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::connect):
   - Pass the isAutomaticInspection and immediatelyPause flags.

(WebKit::ServiceWorkerDebuggableProxy::pauseWaitingForAutomaticInspection):
(WebKit::ServiceWorkerDebuggableProxy::unpauseForResolvedAutomaticInspection):
   - Customize pausing behavior for ServiceWorkerDebuggableProxy.
   - The m_isPausedWaitingForAutomaticInspection is repurposed to
     represent if pausing was requested, and it&apos;s never marked to false
     again because that&apos;d need a message from the web process upon
     unpausing which is not worth the effort for now.

(WebKit::ServiceWorkerDebuggableProxy::disconnect):
(WebKit::ServiceWorkerDebuggableProxy::dispatchMessageFromRemote):
   - Following pauseWaitingForAutomaticInspection, make logging more
     descriptive.

* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::frontendInitialized):
   - In step 4, unblock the paused service worker thread.

(WebCore::WorkerInspectorController::connectFrontend):
(WebCore::WorkerInspectorController::disconnectFrontend):
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
   - The unpausing logic turned out to work for both the ServiceWorkerDebuggableProxy
     and the old debuggable, so we can remove needing the m_frontendInitializedCallback.

* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
   - Don&apos;t support auto-inspecting the old debuggable when the new proxy
     is used.

Canonical link: <a href="https://commits.webkit.org/292318@main">https://commits.webkit.org/292318@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/04a67ab8b5c1b7856afe3ddbfae52303d56c6bcc

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/95469 "Build is in progress. Recent messages:OS: Sequoia (15.3), Xcode: 16.2; Running apply-patch; Checked out pull request; check-webkit-style") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/131/builds/15071 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/138/builds/4950 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/100517 "Build is in progress. Recent messages:Printed configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 win](https://ews-build.webkit.org/#/builders/59/builds/45974 "Build is in progress. Recent messages:Printed configuration") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/97513 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/130/builds/15356 "Built successfully") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/123/builds/23502 "Build is in progress. Recent messages:OS: Sequoia (15.3), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Compiled WebKit (warnings)") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/72817 "Passed tests") | [✅ 🧪 win-tests](https://ews-build.webkit.org/#/builders/60/builds/30082 "Passed tests") 
| [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/98472 "Build is in progress. Recent messages:Running configuration; Skipping applying patch since patch_id isn't provided; Checked out pull request; Passed webkitperl tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/132/builds/11504 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/86225 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/53149 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/133/builds/11210 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/135/builds/3935 "Passed tests") | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/45310 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/88137 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/81404 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/136/builds/4054 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/102553 "Built successfully") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/94089 "Built successfully and passed tests") | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/128/builds/22519 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/122/builds/16449 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/81854 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/121/builds/22771 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/82235 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/81205 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/20372 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/126/builds/25788 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/137/builds/3250 "Passed tests") | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 playstation](https://ews-build.webkit.org/#/builders/134/builds/15846 "Build is in progress. Recent messages:Printed configuration") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv](https://ews-build.webkit.org/#/builders/127/builds/22487 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Checked out pull request; Running compile-webkit") | [✅ 🛠 mac-safer-cpp](https://ews-build.webkit.org/#/builders/120/builds/27641 "Built successfully") | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/116882 "Built successfully") | 
| | [![loading](https://user-images.githubusercontent.com/3098702/171232313-daa606f1-8fd6-4b0f-a20b-2cb93c43d19b.png) 🛠 tv-sim](https://ews-build.webkit.org/#/builders/125/builds/22146 "Build is in progress. Recent messages:OS: Sequoia (15.2), Xcode: 16.2; Skipping applying patch since patch_id isn't provided; Running checkout-pull-request; Running compile-webkit") | | [  ~~🧪 jsc-armv7-tests~~](https://ews-build.webkit.org/#/builders/25/builds/33336 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/129/builds/25622 "Built successfully") | | | 
| | [⏳ 🛠 watch-sim ](https://ews-build.webkit.org/#/builders/watchOS-11-Simulator-Build-EWS "Waiting in queue, processing has not started yet") | | | 
<!--EWS-Status-Bubble-End-->

burg commented on Source/WebKit/UIProcess/WebProcessProxy.messages.in:
```diff
@@ -102,8 +102,8 @@ messages -> WebProcessProxy WantsDispatchMessage {
 
 #if ENABLE(REMOTE_INSPECTOR) && PLATFORM(COCOA)
-    CreateServiceWorkerDebuggable(WebCore::ServiceWorkerIdentifier identifier, URL url)
+    CreateServiceWorkerDebuggable(WebCore::ServiceWorkerIdentifier identifier, URL url, enum:bool WebCore::ServiceWorkerIsInspectable isInspectable) -> (bool shouldWaitForAutoInspection) Synchronous
```
Do we actually expect/need this to be synchronous if the debuggable thread is going to spin and wait anyway?

the-chenergy replied:
This actually works either being synchronous or async. From the rest of [`WebProcessProxy.messages.in`](https://github.com/WebKit/WebKit/blob/c160b11edda887fab8467eb9e37cf06e68d95fb2/Source/WebKit/UIProcess/WebProcessProxy.messages.in) it looked like we have preferred synchronous replies, and it avoided the need for another completion callback.

Did you think async would make more sense though?

burg replied:
I'm not sure-in the past we avoided Sync IPC at all costs. cc @achristensen07 @cdumez for comment.

cdumez replied:
We should always avoid sync IPC whenever possible.

burg commented on Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
```diff
@@ -74,6 +82,23 @@ void ServiceWorkerDebuggableProxy::dispatchMessageFromRemote(String&& message)
+
+void ServiceWorkerDebuggableProxy::unpauseForResolvedAutomaticInspection()
+{
+    RELEASE_LOG(Inspector, "ServiceWorkerDebuggableProxy::unpauseForResolvedAutomaticInspection");
```
I'm not sure we want this to be a RELEASE_LOG, or at least maybe it could be a more readable message.

the-chenergy replied:
I was blindly following the rest of that [`ServiceWorkerDebuggableProxy.cpp`](https://github.com/WebKit/WebKit/blob/c160b11edda887fab8467eb9e37cf06e68d95fb2/Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp) file. Although I do feel like this could be more readable or at least log something object-identifiable?

burg reviewed (approved):
Please mention testing with and without Develop menu enabled. This was why we had to roll it out originally.

the-chenergy committed:
Web Inspector: The new ServiceWorkerDebuggableProxy needs to adopt auto inspection
rdar://145795990
https://bugs.webkit.org/show_bug.cgi?id=288776

Reviewed by BJ Burg.

This work adds support to automatically inspect or pause the
ServiceWorkerDebuggableProxy, alongside minor refactoring of existing
code around auto-inspection.

The squashed commit includes three patches:

(Patch 1 of 3) Make pausing for automatic inspection a responsibility
of the RemoteInspectionTarget rather than RemoteInspector.

To summarize how automatic inspection generally works:
1. The web content marks its debuggable as inspectable.
2. The debuggable blocks its thread before starting code execution.
3. The auto-launched frontend, after initializing, notifies the backend.
4. The backend unblocks the debuggable to resume its code execution.

As the newly introduced ServiceWorkerDebuggableProxy lives in the UI
process but controls a different thread, during step 2, it's the service
worker thread that would need to be blocked, while the UI process
remains free to handle incoming messages from the remote inspector host.

This patch builds the ground work for supporting pausing in a different
thread by moving the handling of pausing or unpausing into the
debuggable instead of implicitly in RemoteInspector. With this change:
- From the RemoteInspector's perspective, it only needs to manage the
  set of debuggables waiting for a reply of auto-inspection policy.
  Notably, it now tells the debuggable to unpause when auto-inspection
  is rejected, rather than relying on the call to pauseWaitingForAutomaticInspection
  to return and assuming that it was a blocking operation.
- From the RemoteInspectionTarget's perspective, it should decide
  what the pausing behavior is, defaulting to blocking the current
  thread with a busy loop. The pause should then persist until an
  explicit call of unpauseForResolvedAutomaticInspection.

With this patch, it'll be easier to let ServiceWorkerDebuggableProxy
later define a custom pausing behavior that controls a different thread.

I tested on both Mac and iOS, with an app that spawns a JSContext, that
this change preserves normal and automatic inspection with no noticeable
difference from before.

* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspectionTarget.cpp:
(Inspector::RemoteInspectionTarget::pauseWaitingForAutomaticInspection):
(Inspector::RemoteInspectionTarget::unpauseForResolvedAutomaticInspection):
(Inspector::RemoteInspectionTarget::unpauseForInitializedInspector): Deleted.
   - Use m_isPausedWaitingForAutomaticInspection to explicitly handle
     the period of pausing.
   - Rename unpauseForResolvedAutomaticInspection to indicate that this
     should be called in both the frontend-initialized case and, as of
     now, the rejected case as well.

* Source/JavaScriptCore/inspector/remote/RemoteConnectionToTarget.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteConnectionToTargetCocoa.mm:
(Inspector::RemoteConnectionToTarget::setup):
   - Make sure to call setupCompleted when we've determined the answer
     of isAutomaticInspection to clean up that hash set entry, since
     it's no longer the debuggable's job to do this upon
     frontend-initialized.

* Source/JavaScriptCore/inspector/remote/RemoteInspector.h:
* Source/JavaScriptCore/inspector/remote/RemoteInspector.cpp:
* Source/JavaScriptCore/inspector/remote/cocoa/RemoteInspectorCocoa.mm:
   - Rename m_automaticInspectionCandidates since "pausing" is no longer
     this object's responsibility, and this hash set is only here to
     know which incoming inspection requests were automatic.

(Inspector::RemoteInspector::waitingForAutomaticInspection): Deleted.
   - No longer used. Debuggables should not rely on this information
     to remain paused.

(Inspector::RemoteInspector::setupFailed):
(Inspector::RemoteInspector::setupCompleted):
(Inspector::RemoteInspector::setPendingMainThreadInitialization):
(Inspector::RemoteInspector::updateAutomaticInspectionCandidate):
(Inspector::RemoteInspector::sendAutomaticInspectionCandidateMessage):
(Inspector::RemoteInspector::stopInternal):
(Inspector::RemoteInspector::setupXPCConnectionIfNeeded):
(Inspector::RemoteInspector::xpcConnectionFailed):
(Inspector::RemoteInspector::receivedSetupMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionConfigurationMessage):
(Inspector::RemoteInspector::receivedAutomaticInspectionRejectMessage):
* Source/JavaScriptCore/inspector/remote/glib/RemoteInspectorGlib.cpp:
(Inspector::RemoteInspector::stopInternal):
* Source/JavaScriptCore/inspector/remote/socket/RemoteInspectorSocket.cpp:
(Inspector::RemoteInspector::stopInternal):
* Source/WebCore/inspector/InspectorController.cpp:
(WebCore::InspectorController::frontendInitialized):
* Source/JavaScriptCore/inspector/JSGlobalObjectInspectorController.cpp:
(Inspector::JSGlobalObjectInspectorController::frontendInitialized):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
   - Adapt to renamed things.

---

(Patch 2 of 3) Move setup of ServiceWorkerDebuggable auto-inspection
into WebSWContextManagerConnection.

Implementing automatic inspection of the ServiceWorkerDebuggable
required marking the debuggable as inspectable and then handling when
to unpause. That work was originally done inside ServiceWorkerThreadProxy.
As we plan to implement the same function for ServiceWorkerDebuggableProxy,
which is created and largely controlled in WebSWContextManagerConnection,
it might be more readable and logical to have both implementations in
that class instead.

* Source/WebCore/workers/service/context/SWContextManager.cpp:
(WebCore::SWContextManager::registerServiceWorkerThreadForInstall):
(WebCore::SWContextManager::stopRunningDebuggerTasksOnServiceWorker):
* Source/WebCore/workers/service/context/SWContextManager.h:
(WebCore::SWContextManager::registerServiceWorkerThreadForInstall):
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.cpp:
(WebCore::ServiceWorkerThreadProxy::ServiceWorkerThreadProxy):
(WebCore::ServiceWorkerThreadProxy::threadStartedRunningDebuggerTasks): Deleted.
* Source/WebCore/workers/service/context/ServiceWorkerThreadProxy.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
   - Move the work of handling the debuggable (steps 2 and 4) into WebSWContextManagerConnection
     where more implementation is anticipated.
   - Lay down a basic code structure for handling auto-inspection for
     when either the Debuggable or the DebuggableProxy is used.

* Source/WTF/wtf/PlatformEnableCocoa.h:
   - Add back the feature flag here as my work should end up providing
     support for both ServiceWorkerDebuggable and ServiceWorkerDebuggableProxy.

---

(Patch 3 of 3) Add support for auto-inspecting ServiceWorkerDebuggableProxy.

This patch adds support to automatically inspect or pause a
ServiceWorkerDebuggableProxy in a similar fashion to the same features
for JSGlobalObjectInspectorController and ServiceWorkerDebuggable.
The ServiceWorkerDebuggableProxy is unique in that, in step 2, it should
keep the UI process unblocked to handle incoming remote inspector
messages including the auto-inspection policy, while controlling when
to unblock the actual service worker thread. This is done by two IPC
messages:
- The async callback of CreateServiceWorkerDebuggable, which can
  indicate to unpause because there's no eligible remote inspector
  at all.
- The new message UnpauseServiceWorkerForRejectedAutomaticInspection can
  request unpause when the policy is received and it's a rejection.

These messages are from the UI process into the web process so they
should be relatively safe. If neither of the above two cases are met,
that means auto-inspection was accepted, and the web process will listen
to the frontendInitialized event to unblock the service worker thread.

I tested on Mac and iOS with a inspector host having Develop menu
on or off, or without support of auto-inspecting service workers and
didn't experience problems or hangs like what my last attempt introduced.

* Source/WebKit/UIProcess/WebProcessProxy.h:
* Source/WebKit/UIProcess/WebProcessProxy.messages.in:
* Source/WebKit/UIProcess/Cocoa/WebProcessProxyCocoa.mm:
(WebKit::WebProcessProxy::createServiceWorkerDebuggable):
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.h:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.messages.in:
* Source/WebKit/WebProcess/Storage/WebSWContextManagerConnection.cpp:
(WebKit::WebSWContextManagerConnection::installServiceWorker):
(WebKit::WebSWContextManagerConnection::unpauseServiceWorkerForRejectedAutomaticInspection):
   - Step 1 and 2, creating and blocking the thread as needed.

(WebKit::WebSWContextManagerConnection::connectToInspector):
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.h:
* Source/WebCore/workers/service/context/ServiceWorkerInspectorProxy.cpp:
(WebCore::ServiceWorkerInspectorProxy::connectToWorker):
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.h:
* Source/WebKit/WebProcess/Inspector/ServiceWorkerDebuggableProxy.cpp:
(WebKit::ServiceWorkerDebuggableProxy::connect):
   - Pass the isAutomaticInspection and immediatelyPause flags.

(WebKit::ServiceWorkerDebuggableProxy::pauseWaitingForAutomaticInspection):
(WebKit::ServiceWorkerDebuggableProxy::unpauseForResolvedAutomaticInspection):
   - Customize pausing behavior for ServiceWorkerDebuggableProxy.
   - The m_isPausedWaitingForAutomaticInspection is repurposed to
     represent if pausing was requested, and it's never marked to false
     again because that'd need a message from the web process upon
     unpausing which is not worth the effort for now.

(WebKit::ServiceWorkerDebuggableProxy::disconnect):
(WebKit::ServiceWorkerDebuggableProxy::dispatchMessageFromRemote):
   - Following pauseWaitingForAutomaticInspection, make logging more
     descriptive.

* Source/WebCore/inspector/WorkerInspectorController.cpp:
(WebCore::WorkerInspectorController::frontendInitialized):
   - In step 4, unblock the paused service worker thread.

(WebCore::WorkerInspectorController::connectFrontend):
(WebCore::WorkerInspectorController::disconnectFrontend):
* Source/WebCore/inspector/WorkerInspectorController.h:
* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.cpp:
(WebCore::ServiceWorkerDebuggable::connect):
   - The unpausing logic turned out to work for both the ServiceWorkerDebuggableProxy
     and the old debuggable, so we can remove needing the m_frontendInitializedCallback.

* Source/WebCore/workers/service/context/ServiceWorkerDebuggable.h:
   - Don't support auto-inspecting the old debuggable when the new proxy
     is used.

Canonical link: https://commits.webkit.org/292318@main
